### ISSUE TITLE
Different results between model.evaluate() and model.predict()

### ISSUE THREAD

**Issue by parkerzf** [fchollet/keras#5140]
I got different results between model.evaluate() and model.predict(). Could someone point out what is wrong in my calculation as follows? Note that the **model, X_test_features, y_regression_test** are identical in two approaches.

Thank you very much!

1. directly use model evaluate() to get loss and metrics:
```python
model = define_top_model()
model.compile(loss='mse', optimizer='rmsprop', metrics=['mae', 'mape'])
model.load_weights(model_weights_file)
scores = model.evaluate(X_test_features, y_regression_test, batch_size=batch_size)
logger.info('mse=%f, mae=%f, mape=%f' % (scores[0],scores[1],scores[2]))
```
The output is:
**mse=0.551147, mae=0.589529, mape=10.979756**

2. get the preds numpy array using model.predict(), and use keras metrics to calculate metrics:
```python
model = define_top_model()
model.compile(loss='mse', optimizer='rmsprop', metrics=['mae', 'mape'])
model.load_weights(model_weights_file)
preds = model.predict(X_test_features, batch_size=batch_size)
tf_session = K.get_session()
mse = metrics.mean_squared_error(y_regression_test, preds)
mae = metrics.mean_absolute_error(y_regression_test, preds)
mape = metrics.mean_absolute_percentage_error(y_regression_test, preds)
logger.info('mse=%f, mae=%f, mape=%f' % (mse.eval(session=tf_session),
                                                             mae.eval(session=tf_session),
                                                             mape.eval(session=tf_session)))
```

The output is:
**mse=0.678286, mae=0.654362, mape=12.249291**

**Comment by bstriner** [comment]
Weird. Do you have a full script that can reproduce this? Maybe the y_regression_test has a weird dtype or shape that is being cleaned by Keras but not when passed to TF directly. predict and evaluate should both be setting training mode to 0 so there shouldn't be a difference. Are there any strange layers or something else that might explain the issue?

I'm wondering whether the issue is different predictions or different metrics calculations. If you do this, which version does it match? If it matches 1, then your metric calculation is different than keras. If it matches 2, then your predictions are different than keras.

```python
x = Input(preds.shape[1:])
m = Model(x,x)
m.compile(loss='mse', optimizer='rmsprop', metrics=['mae', 'mape'])
m.evaluate(preds, y_regression_test, batch_size=batch_size)
```

Cheers

**Comment by parkerzf** [comment]
@bstriner Thank you very much! It's a good idea. I will check it tomorrow.

**Comment by parkerzf** [comment]
@bstriner I did the experiment you suggested. 
Here is the updated model outputs:

output 1: mse=0.511590, mae=0.565474, mape=11.014793
output 2: mse=0.581871, mae=0.602397, mape=11.728496

The identity model output is the same result as output 1, which means that I calculate the metrics different to the evaluate function.

It seems very strange to me. So I did one more experiment. I use numpy function to calculate mse, however, the result is the same as output 2, i.e. mse=0.581871

```python
logger.info('mse=%f' % np.mean(np.square(y_regression_test - preds)))
```

Do you know what are the possible reason for this?

I will upload the script and the preds and y_regression files so that you can repeat the result if you want.

**Comment by parkerzf** [comment]
Here is the script to repeat my experiment:

https://github.com/parkerzf/keras_metrics

The outputs are:
manually result: mse=0.581871
evaluate result: mse=0.511590

**Comment by bstriner** [comment]
@parkerzf The issue is that your y is the wrong shape and dtype, but keras automatically fixes the shape for you, giving different results if you use the model or don't.

This line fixes the issue:
```python
y_regression = y_regression.astype(np.float32).reshape((-1,1))
```

This script should have the same values through numpy, evaluate, or using the backend.

```python
import numpy as np
from keras.models import Model
from keras.layers import Input
from keras.metrics import mse, mae, mape
from keras import backend as K

def test(preds, y_regression):
	print("preds: {}, {}".format(preds.dtype, preds.shape))
	print("y_regression: {}, {}".format(y_regression.dtype, y_regression.shape))

	print('manual result: mse=%f' % np.mean(np.square(y_regression - preds)))

	a = mse(y_regression, preds)
	b = mae(y_regression, preds)
	c = mape(y_regression, preds)
	f = K.function([], [a,b,c])
	print 'backend result: mse={}, mae={}, mape={}'.format(*f([]))

	x = Input(preds.shape[1:])
	m = Model(x, x)
	m.compile(loss='mse', optimizer='rmsprop', metrics=['mae', 'mape'])
	scores = m.evaluate(preds, y_regression, batch_size=32, verbose=0)

	print '\nevaluate result: mse={}, mae={}, mape={}'.format(*scores)

preds = np.load(open('preds.npy','rb'))
y_regression = np.load(open('y_regression.npy','rb'))
print("Before fix")
test(preds, y_regression)
#this line fixes the issue
print("Fixing dtype and shape")
y_regression = y_regression.astype(np.float32).reshape((-1,1))
print("After fix")
test(preds, y_regression)
```

> Before fix
> preds: float32, (7666L, 1L)
> y_regression: float64, (7666L,)
> manual result: mse=0.581871
> backend result: mse=0.581870887686, mae=0.602397143873, mape=11.7284955271
> evaluate result: mse=0.511590141619, mae=0.565474303248, mape=11.0147933211
> Fixing dtype and shape
> 
> After fix
> preds: float32, (7666L, 1L)
> y_regression: float32, (7666L, 1L)
> manual result: mse=0.511590
> backend result: mse=0.511590123177, mae=0.565474271774, mape=11.014793396
> evaluate result: mse=0.511590141619, mae=0.565474303248, mape=11.0147933211
> 

Cheers

**Comment by parkerzf** [comment]
Thank you so much!

**Comment by bstriner** [comment]
No problem! Please close the issue if all is well and good.

Cheers

**Comment by jybbjybb** [comment]
I found the same thing, model.predict() and model.evaluate() do not match.
Here is part of the sample code on cifar10 dataset:
``
    y_pred = model.predict(x_test,batch_size=100,verbose=1)
    Y_pred = np.argmax(y_pred,axis=1)
    Y_test = np.argmax(y_test,axis=1)
    accuracy = (len(Y_test) - np.count_nonzero(Y_pred - Y_test) + 0.0)/len(Y_test)
    score = model.evaluate(x_test, y_test, batch_size=100, verbose=1)
``
accuracy is 0.8991 but score[1] is 0.9804

**Comment by jybbjybb** [comment]
I found the same thing, model.predict() and model.evaluate() do not match.
Here is part of the sample code on cifar10 dataset:
#####################################################
y_pred = model.predict(x_test,batch_size=100,verbose=1) 
Y_pred = np.argmax(y_pred,axis=1) 
Y_test = np.argmax(y_test,axis=1) 
accuracy = (len(Y_test) - np.count_nonzero(Y_pred - Y_test) + 0.0)/len(Y_test) 
score = model.evaluate(x_test, y_test, batch_size=100, verbose=1)
####################################################
accuracy is 0.8991 but score[1] is 0.9804

**Comment by chaoyan1073** [comment]
I found the similar problem, however, I got the same accuracy while different loss.
my code is like:
 ```
        test_logloss, test_acc = model.evaluate(X_test, Y_test, verbose=0, batch_size=64)
        print("My test loss: ", test_logloss)
        test_predict = model.predict(X_test, verbose=0, batch_size=batch_size)
        test_predict_tf = tf.convert_to_tensor(test_predict, 'float32')
        print("true: {}, {}".format(Y_test.dtype, Y_test.shape))
        print("preds: {}, {}".format(test_predict.dtype, test_predict.shape))
        test_loss2 = keras.losses.categorical_crossentropy(Y_test, test_predict_tf)
        test_loss2 = tf.Session().run(test_loss2)
        test_loss2 = np.mean(test_loss2)
        print("My test loss2: ", test_loss2)
```

**Comment by newbiestatguy** [comment]
model.evaluate is not the function that tells you how accurately your model is going to classify in the actual world if you put it into production. For that you need to write your own code.

**Comment by kev5** [comment]
How do I plot the confusion matrix in my code?

```
from getEmbeddings import getEmbeddings
import matplotlib.pyplot as plt
import numpy as np
import keras
from keras import backend as K
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers import Dense, Dropout, LSTM, Embedding, Input, RepeatVector
from keras.optimizers import SGD
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
import scikitplot.plotters as skplt


def plot_cmat(yte, ypred):
    '''Plotting confusion matrix'''
    skplt.plot_confusion_matrix(yte,ypred)
    plt.show()

xtr,xte,ytr,yte = getEmbeddings("datasets/train.csv")
np.save('./xtr',xtr)
np.save('./xte',xte)
np.save('./ytr',ytr)
np.save('./yte',yte)

xtr = np.load('./xtr.npy')
xte = np.load('./xte.npy')
ytr = np.load('./ytr.npy')
yte = np.load('./yte.npy')


def baseline_model():
    '''Neural network with 3 hidden layers'''
    model = Sequential()
    model.add(Dense(256, input_dim=300, activation='relu', kernel_initializer='normal'))
    model.add(Dropout(0.3))
    model.add(Dense(256, activation='relu', kernel_initializer='normal'))
    model.add(Dropout(0.5))
    model.add(Dense(80, activation='relu', kernel_initializer='normal'))
    model.add(Dense(2, activation="softmax", kernel_initializer='normal'))

    # gradient descent
    sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
    
    # not 100% sure what "compile" does, maybe just runs the gradient descent algorithm
    model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])
    return model


model = baseline_model()
model.summary()
x_train, x_test, y_train, y_test = train_test_split(xtr, ytr, test_size=0.2, random_state=42)
label_encoder = LabelEncoder()
label_encoder.fit(y_train)
encoded_y = np_utils.to_categorical((label_encoder.transform(y_train)))
label_encoder.fit(y_test)
encoded_y_test = np_utils.to_categorical((label_encoder.transform(y_test)))
estimator = model.fit(x_train, encoded_y, epochs=20, batch_size=64)
print("Model Trained!")
score = model.evaluate(x_test, encoded_y_test)
print("")
print("Accuracy = " + format(score[1]*100, '.2f') + "%")   # 92.62%
```

**Comment by sinberlin2** [comment]
Hi I had the same issue. I fixed it by setting the batch size in model.evaluate to the length of the predictions!

**Comment by mallisha** [comment]
Even i hve the same issue my model.evaluate give me an accuracy of 89 whereas my model.predict about 40 any idea how to solve to?? Tried lots of thinge stuck here

**Comment by sonukiller** [comment]
If you are using `tf.keras.preprocessing.image.ImageDataGenerator` for loading data, try setting `shuffle=False` for test data.
